sys/kern/cred_syscalls.c

```diff
@@ -131,3 +139,110 @@ int do_setgroups(proc_t *p, int ngroups, const gid_t *gidset) {
 
   return 0;
 }
+
+int do_setuid(proc_t *p, uid_t uid) {
+  proc_lock(p);
+
+  uid_t ruid = uid, euid = uid, suid = uid;
+
+  if (p->p_cred.cr_euid == 0)
+    ruid = suid = -1;
+
+  int error = change_resuid(&p->p_cred, ruid, euid, suid);
+
+  proc_unlock(p);
+  return error;
+}
+
+int do_seteuid(proc_t *p, uid_t euid) {
+  proc_lock(p);
+  int error = change_resuid(&p->p_cred, -1, euid, -1);
+  proc_unlock(p);
+  return error;
+}
+
+int do_setreuid(proc_t *p, uid_t ruid, uid_t euid) {
+  int error;
+  proc_lock(p);
+
+  uid_t cur_euid = p->p_cred.cr_euid;
+  uid_t cur_ruid = p->p_cred.cr_ruid;
+  uid_t cur_suid = p->p_cred.cr_suid;
+
+  /* if proc is not root it can set ruid only to euid */
+  if (cur_euid != 0 && ruid != (uid_t)-1 && ruid != cur_euid) {
+    error = EPERM;
+    goto fail;
+  }
+
+  /* if ruid is set
+   * or euid is set to value other than current suid or ruid(and of course euid)
+   * then suid is set to current euid
+   *
+   * when we don't set ruid and euid we also don't set suid
+   * (the last condition in or)
+   */
+  if (ruid == (uid_t)-1 && (euid == cur_ruid || euid == cur_euid ||
+                            euid == cur_suid || euid == (uid_t)-1))
+    cur_euid = -1;
+
+  error = change_resuid(&p->p_cred, ruid, euid, cur_euid);
+
+fail:
```

Comment:
`end` feels like a more appropriate name for this label, since we arrive here even if we succeed.

Reply:
Yes. `end` will be beter.